proxy/src/outbound.rs

```diff
@@ -51,81 +51,78 @@ impl<B> Outbound<B> {
 {
     type Request = http::Request<B>;
     type Response = bind::HttpResponse;
     type Error = <Self::Service as tower::Service>::Error;
-    type Key = (Destination, Protocol);
+    type Key = Dst;
     type RouteError = bind::BufferSpawnError;
     type Service = InFlightLimit<Timeout<Buffer<Balance<
         load::WithPendingRequests<Discovery<B>>,
         choose::PowerOfTwoChoices<rand::ThreadRng>
     >>>>;
 
+    /// Determines the destination service for a given Request.
+    ///
+    /// Destinations are either bound or unbound. Unbound destinations must be resolved
+    /// through the controller to produce a list of endpoints. Bound destinations use the
+    /// client-provided destination address.
+    ///
+    /// For HTTP/2 requests, a service may support several :authority values. HTTP/1
+    /// requests, on the other hand, may not share a service across authorities.
+    ///
+    /// Services may not be reused if they do not include an authority.
     fn recognize(&self, req: &Self::Request) -> Option<Reuse<Self::Key>> {
         let proto = bind::Protocol::detect(req);
 
-        // The request URI and Host: header have not yet been normalized
-        // by `NormalizeUri`, as we need to know whether the request will
-        // be routed by Host/authority or by SO_ORIGINAL_DST, in order to
-        // determine whether the service is reusable.
-        let local = req.uri().authority_part()
+        // Determine the request's `:authority` (or `Host`), either from its headers or
+        // from the request's URI.
+        //
+        // The authority is normalized to determine whether or not it should be resolved
+        // through the controller.
+        let qualified_name = req.uri().authority_part()
             .cloned()
-        // Therefore, we need to check the host header as well as the URI
-        // for a valid authority, before we fall back to SO_ORIGINAL_DST.
             .or_else(|| h1::authority_from_host(req))
             .map(|authority| {
-                FullyQualifiedAuthority::normalize(
-                    &authority,
-                    &self.default_namespace)
+                FullyQualifiedAuthority::normalize(&authority, &self.default_namespace)
             });
 
-        // If we can't fully qualify the authority as a local service,
-        // and there is no original dst, then we have nothing! In that
-        // case, we return `None`, which results an "unrecognized" error.
+        // If the request has an authority that looks like something the destination
+        // service might be able to resolve, treat the name as unbound. Note that this
+        // will result in different keys for different original authority values, so
+        // requests for `foo` and `foo.ns` won't be satisfied by the same connections.
+        if let Some(qn) = qualified_name {
+            if qn.use_destination_service {
+                return Some(Reuse::Reusable(Dst::Unbound(qn.name, proto)));
```

Comment:
This may be an unpopular opinion, but I find that early returns tend to make functions like this less clear.

Reply:
Moving logic into else clauses would make this even more convoluted. Early returns are a necessary evil.